### ISSUE TITLE
Attempting to check out a non-existent commit stores the bad SHA-1 in HEAD

### ISSUE THREAD

**Issue by magnusbaeck** [go-git/go-git#457]
If you attempt to check out a non-existent commit SHA-1 that operation fails, unsurprisingly, with plumbing.ErrObjectNotFound. What's surprising is that the SHA-1 remains in .git/HEAD. This isn't consistent with cgit's behavior and can result in some really confusing errors. Here's a minimal example:

```go
package main

import (
	"path/filepath"

	"github.com/go-git/go-billy/v5/osfs"
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/cache"
	"github.com/go-git/go-git/v5/storage/filesystem"
)

const gitPath = "./git"

func main() {
	worktreeFS := osfs.New(gitPath)
	gitDirFS := osfs.New(filepath.Join(gitPath, ".git"))
	repo, err := git.Init(filesystem.NewStorage(gitDirFS, cache.NewObjectLRUDefault()), worktreeFS)
	if err != nil {
		panic(err)
	}
	tree, err := repo.Worktree()
	if err != nil {
		panic(err)
	}
	err = tree.Checkout(
		&git.CheckoutOptions{
			Hash:  plumbing.NewHash("1dafc1f4067af82145a8fb8248b11df5634dc3e8"), // Any random SHA-1
			Force: true,
		})
	if err != nil {
		panic(err)
	}
}
```

Result after running this:

```
$ cd git
$ cat .git/HEAD 
1dafc1f4067af82145a8fb8248b11df5634dc3e8
$ git status
fatal: bad object HEAD
```

I discovered this when writing code that speculatively checked out a commit that _might_ be present locally. When that failed with plumbing.ErrObjectNotFound I fetched a ref that I knew pointed to the desired commit. However, because HEAD pointed to the commit in question the SHA-1 was listed not only in the _wants_ set of the fetch operation but also the _haves_ set, so the fetch operation failed with transport.ErrEmptyUploadPackRequest.

It looks like this would be fairly easy to address by simply verifying that opts.Hash resolves to something that can be checked out prior to this code in Worktree.Checkout:

https://github.com/go-git/go-git/blob/f92011d95f98f5deea4959c7d432704a4300d3a8/worktree.go#L175-L179

**Comment by onee-only** [comment]
I think it was fixed in #966.
Tested with latest release and it works as expected.

**Comment by pjbgf** [comment]
@onee-only nice one, thank you.

Closing as per above.
